api/proquest/importer.py

```diff
@@ -0,0 +1,554 @@
+
+        :param patron: Patron object
+        :type patron: core.model.patron.Patron
+
+        :param configuration: Configuration object
+        :type configuration: ProQuestOPDS2ImporterConfiguration
+
+        :return: ProQuest JWT bearer token
+        :rtype: str
+        """
+        affiliation_id = self._get_patron_affiliation_id(patron, configuration)
+
+        try:
+            token = self._api_client.create_token(self._db, affiliation_id)
+            token_expiration_timeout = (
+                int(configuration.token_expiration_timeout)
+                if configuration.token_expiration_timeout
+                else ProQuestOPDS2ImporterConfiguration.DEFAULT_TOKEN_EXPIRATION_TIMEOUT_SECONDS
+            )
+
+            self._credential_manager.save_proquest_token(
+                self._db,
+                patron,
+                datetime.timedelta(seconds=token_expiration_timeout),
+                token,
+            )
+
+            return token
+        except Exception as exception:
+            raise CannotCreateProQuestTokenError(exception)
+
+    def _get_or_create_proquest_token(self, patron, configuration):
+        """Get an existing or create a new ProQuest JWT bearer token.
+
+        :param patron: Patron object
+        :type patron: core.model.patron.Patron
+
+        :param configuration: Configuration object
+        :type configuration: ProQuestOPDS2ImporterConfiguration
+
+        :return: ProQuest JWT bearer token
+        :rtype: str
+        """
+        token = self._credential_manager.lookup_proquest_token(self._db, patron)
+
+        self._logger.info(
+            "Patron {0} has the following token: {1}".format(patron, token)
+        )
+
+        if not token:
+            token = self._create_proquest_token(patron, configuration)
+
+        return token
+
+    def _get_book(self, patron, configuration, document_id):
+        """Get a book's content (in the case of open-access books) or a book's link otherwise.
+
+        :param patron: Patron object
+        :type patron: core.model.patron.Patron
+
+        :param configuration: Configuration object
+        :type configuration: ProQuestOPDS2ImporterConfiguration
+
+        :param document_id: ProQuest Doc ID
+        :type document_id: str
+
+        :return: Either an ACS link to the book or the book content
+        :rtype: Union[str, bytes]
+        """
+        token = self._get_or_create_proquest_token(patron, configuration)
+        iterations = 0
+
+        while True:
+            try:
+                book = self._api_client.get_book(self._db, token, document_id)
+
+                return book
+            except HTTPError as exception:
+                if exception.response.status_code != 401 or iterations >= 1:
+                    raise
+                else:
+                    token = self._create_proquest_token(patron, configuration)
+
+            iterations += 1
+
+    def extract_next_links(self, feed):
```

Comment:
It could work but I would prefer to handle downloading feed pages inside `ProQuestAPIClient._download_feed_page` because it contains additional validation logic. ProQuest API is not very reliable: sometimes they return JSON, sometimes they return some other content (for example, an open access book) so I added a lot of code validating responses we receive from their API